### PR TITLE
perf(dashboard): make oracle chart zoom/pan snappy via rAF-coalesced relayout

### PR DESCRIPTION
**Pull request by chapati23**
## What

The oracle chart (`/pool/[poolId]` → Oracle tab) felt very laggy when zooming/scrolling. Profiling pinned it to **the main thread, not the GPU**: every wheel tick ran a full decimation recompute + `Plotly.react` redraw **twice** (the wheel's own `Plotly.relayout` plus a React `setVisibleRange` round-trip), so a zoom gesture pegged the main thread at ~10 fps.

A DevTools trace on a busy pool showed the cost is **~95% JS scripting, ~1% painting** — so this is a scripting problem. Switching to `scattergl`/WebGL would barely move the needle (paint is already ~1%) and would cost a Plotly bundle-size bump against a tight `size-limit` budget, so it's explicitly **not** done here.

## How (three bundle-free fixes)

1. **rAF-coalesce the relayout round-trip** (`useCoalescedRelayout`) — collapse the per-tick `onRelayout → setVisibleRange/setShowAll` into **one React update per animation frame**. The wheel's own `Plotly.relayout` still reframes the axis instantly every tick (the zoom is visible immediately); only the React-side work (decimation re-scope, daily-mode re-eval, look-ahead) is deferred to once the viewport settles.
2. **Memoize the trace `data` + `layout`** (extracted into `useOracleChartModel`). react-plotly.js 2.6 ref-compares `data`/`layout`/`config` in `componentDidUpdate` and skips `Plotly.react` when all three are unchanged — so a 30 s SWR repoll or a no-op re-render no longer rebuilds + redraws the SVG.
3. **Binary-search `sliceToWindow`'s window bounds** (rows are sorted ASC) → `O(log n + k)` instead of an `O(n)` scan over the tens of thousands of rows that accumulate after scroll-back, on *every* relayout.

## Measured (busy Celo pool, 1000-point default view, 48-wheel gesture, dev build)

| metric | before | after |
|---|---|---|
| main-thread long-tasks | **50 (7.5 s total)** | **0** |
| longest single freeze | 290 ms | < 50 ms |
| median frame | 70 ms (~10 fps) | **17 ms (~60 fps)** |
| gesture wall-time | 10.9 s | 3.8 s |

Numbers are dev-mode + a synthetic gesture, so read them as relative; prod is faster. The transient blank-edge risk from deferring the decimation re-scope is covered by the existing one-span decimation margin (measured worst trace-vs-axis edge gap **0 s** even at 8 ms trackpad cadence).

## Scope / follow-up

This pass is the bundle-free interaction fix only (as scoped). At *extreme* synthetic cadence (8 ms, faster than a real browser coalesces `wheel`) a per-frame SVG relayout floor remains; if buttery 60 fps at that cadence is ever needed, the next levers are a lower point cap or a custom Plotly bundle with `scattergl` (+ a `size-limit` bump) — deliberately deferred.

## Testing

- New `oracle-chart-coalesce` suite: latest-wins-per-frame, single rAF per burst, re-arm after flush, cleanup-cancels-on-unmount. **69 oracle unit tests pass.**
- Chart output unchanged — browser-verified the band tint, green/red markers, line, rangeslider, and that zoom still works; zero console errors.
- Full pre-push gate green (lint, typecheck, test, test:browser, knip, deps, react-doctor diff + 100/100 score, build, size-limit).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Low Risk**
> Dashboard-only performance and interaction plumbing; behavior is locked by existing oracle tests plus new coalesce/decimation edge-case coverage.
> 
> **Overview**
> Makes oracle chart zoom/pan feel snappy by cutting redundant main-thread work during wheel-driven relayouts, without changing chart behavior.
> 
> **`useCoalescedRelayout`** batches Plotly `onRelayout` into at most one React update per animation frame (latest event wins), so decimation, daily-mode, and look-ahead no longer run on every wheel tick while Plotly still reframes the axis immediately. Pending frames are cancelled on unmount and **during render** when `uirevision` / pool identity changes, so a stale pool’s range cannot flush after switching pools.
> 
> **`useOracleChartModel`** memoizes decimated traces, layout, and `data` so react-plotly can skip `Plotly.react` when props are referentially unchanged (e.g. SWR repoll or no-op relayout).
> 
> **`sliceToWindow`** in oracle decimation uses binary search on sorted timestamps instead of scanning the full series on each relayout; tests cover empty-slice fallbacks when the window is entirely before or after the data.
> 
> New **`oracle-chart-coalesce`** unit tests cover coalescing, rAF scheduling, reset vs range ordering, and identity-change cancellation.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8d62d3bf419a6df4f06a13cb6e2d25efd9a7023f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: ui-dashboard/src/components/oracle-chart.tsx [hunk 1]
```diff
@@ -116,6 +120,53 @@ function applyOracleRelayout(
...
+        setShowAll,
+        onVisibleXRangeChange,
+      );
+    }
+  }, [setVisibleRange, setShowAll, onVisibleXRangeChange]);
+  useEffect(
+    () => () => {
+      if (rafIdRef.current !== null) cancelAnimationFrame(rafIdRef.current);
+    },
+    [],
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/components/oracle-chart.tsx]: Fixed in d928101 — good catch. `useCoalescedRelayout` now takes `resetKey` (`${networkId}:${poolId}`) and its cleanup effect is keyed on it, so a pool/network switch cancels the pending frame **and** clears the pending refs (the next event re-arms for the new pool). Confirmed against OracleTab's note that it doesn't remount across a `poolId` change. New test: "cancels a pending frame on chart-identity change, then re-arms for the new pool".

File: ui-dashboard/src/components/__tests__/oracle-chart-coalesce.test.tsx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    act(() =>
+      handler(relayout("2026-05-01T00:00:00Z", "2026-05-08T00:00:00Z")),
+    );
+    act(() => root.unmount());
+    mounted = false;
+
+    act(() => flushFrame());
+    expect(setVisibleRange).not.toHaveBeenCalled();
+  });
+});
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/components/__tests__/oracle-chart-coalesce.test.tsx:193]: Fixed in d928101 — added both orders: "latest action wins when range then reset coalesce in one frame" (→ `setShowAll(true)`, `setVisibleRange(null)`, no look-ahead) and "...when reset then range coalesce in one frame" (→ range wins, look-ahead fires). Used the `resetRelayout = () => ({ "xaxis.autorange": true })` helper you suggested.

File: ui-dashboard/src/lib/oracle-decimation.ts [hunk 3]
```diff
@@ -90,11 +90,45 @@ function sliceToWindow<T>(
...
-  });
+  // `rows` is sorted ASC by timestamp (contract above), so binary-search the
+  // window bounds: O(log n + k) instead of an O(n) scan over what can be tens
+  // of thousands of accumulated rows on every relayout during a zoom gesture.
+  const start = lowerBoundByTs(rows, from, getTimestamp); // first ts >= from
+  const end = upperBoundByTs(rows, to, getTimestamp); // first ts > to
+  const sliced = rows.slice(start, end);
   // Never return empty just because the window sits between two sparse points —
   // fall back to the full series so the chart still renders something.
   return sliced.length > 0 ? sliced : rows;
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/lib/oracle-decimation.ts:101]: Fixed in d928101 — added "falls back when all rows sit to the RIGHT of the window (binary-search end=0)" and "...to the LEFT (start=len)", both asserting the full series is returned. Scenario 1 (window between two sparse points) was already covered by the existing "falls back to the full series when the window holds no points" test.

File: ui-dashboard/src/components/oracle-chart.tsx [hunk 4]
```diff
@@ -116,6 +120,53 @@ function applyOracleRelayout(
...
+    pendingRef.current = null;
+    if (e != null) {
+      applyOracleRelayout(
+        e,
+        setVisibleRange,
+        setShowAll,
+        onVisibleXRangeChange,
+      );
+    }
+  }, [setVisibleRange, setShowAll, onVisibleXRangeChange]);
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/components/oracle-chart.tsx]: Fixed in d928101 — `onVisibleXRangeChange` is now held in a ref (refreshed each render) and read at flush time, so a queued frame never fires a stale closure if the prop changes (e.g. `oldestLoadedTs` advances) between schedule and fire. As you noted, this also drops it from `flush`'s deps, so `flush` and the returned handler stay referentially stable across parent re-renders.

File: ui-dashboard/src/components/oracle-chart.tsx [hunk 5]
```diff
@@ -1,7 +1,7 @@
 "use client";
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/components/oracle-chart.tsx:1]: Won't fix (under the enforced cap): the file is 1047 raw lines, but `max-lines` is configured `{ max: 1000, skipBlankLines: true, skipComments: true }` in `eslint.config.mjs` and lint passes — the counted code lines are under 1000; the extra is the explanatory comments this change added. A further split is reasonable hygiene, but extracting the plot-data builders cleanly would need a larger reorg to avoid a circular import (`useOracleChartModel` → `selectOraclePlotData` → `buildOraclePlotData`, which would live in the extracted module and import back), so I'm deferring it rather than folding an unrelated refactor into this perf PR.

**Comment by chapati23** [comment]
@claude — addressed all four notes in d928101:

- **Reset/range coalesce test (P2)** → Fixed: added both orders (range→reset and reset→range), asserting the latest action wins and that reset never fires look-ahead.
- **sliceToWindow fallback untested (P2)** → Fixed: added the all-rows-right (`end=0`) and all-rows-left (`start=len`) binary-search edges; the between-sparse-points case was already covered.
- **Stale `onVisibleXRangeChange` in pending rAF (P3)** → Fixed: held in a ref read at flush time; `flush` and the handler are now referentially stable.
- **File length 1048 lines (P2)** → Won't fix: `max-lines` counts code lines (skipComments + skipBlankLines) and passes under 1000; the raw count is inflated by the added comments. A clean split needs a larger reorg to avoid a circular import — deferred out of this perf PR.

Also addressed Codex's P2: pending relayouts are now cancelled on chart-identity (`resetKey`) change, not just on unmount.

File: ui-dashboard/src/components/oracle-chart.tsx [hunk 6]
```diff
@@ -116,6 +120,73 @@ function applyOracleRelayout(
...
+      );
+    }
+  }, [setVisibleRange, setShowAll]);
+  // Cancel + drop any pending frame on unmount AND on chart-identity change
+  // (`resetKey`). Clearing the refs (not just cancelling) lets the next event
+  // re-arm the rAF for the new pool.
+  useEffect(
+    () => () => {
+      if (rafIdRef.current !== null) cancelAnimationFrame(rafIdRef.current);
+      rafIdRef.current = null;
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/components/oracle-chart.tsx]: Fixed in 8d62d3b — you're right, and good catch. The effect cleanup ran after commit, which is too late against a queued `requestAnimationFrame` (it fires pre-paint). Moved the cancel + clear into a render-phase identity guard (mirroring `useOracleViewport`'s reset on the same `resetKey`), so it runs before the browser's rAF rather than after it. The explicit null-check guard keeps it SSR-safe — `rafIdRef` is always null on the server, so `cancelAnimationFrame` is never called there. Strengthened the test to assert the queued frame is invalidated synchronously on the identity-change render (it now checks the rAF queue has no live callback right after the re-render), instead of relying on `act()` flushing effects.
